Added meilisearch retriever component

MohamedYasserOaf opened the pull request:
for the past 2 weeks , we needed to implement a retriever that utilized meilisearch for hybrid search powered retrieval , the component was created as a custom retriever, you either search an existing index using meilisearch and retrieve context for your LLM or you can upsert an existing or a new index to include new files that can be read using any flowise powered document store or document reading function, then you can retrieve context from this new index, more modifications and improvements may be introduced in future PRs, some commits are related to our personal CI/CD pipeline which can be ignored or re-modified after merge, kindly have a look and address any issues in the comments

HenryHengZJ commented on packages/components/nodes/retrievers/MeilisearchRetriever/MeilisearchRetriever.ts:
```diff
@@ -0,0 +1,177 @@
+        ]
+    }
+    //@ts-ignore
+    vectorStoreMethods = {
```
is this a retriever or a vector store? Because upsert only for vector store

MohamedYasserOaf replied:
well , this component is similar to postgres component , it is essentially a vectorstore (meilisearch index that contain embedded documents) , but also , it is a retriever that utilize hybrid search to retrieve context from the meilisearch index (which was embedded using an external embeddings) , so yes the component does have upsert functionality

HenryHengZJ commented on packages/components/nodes/retrievers/MeilisearchRetriever/MeilisearchRetriever.ts:
```diff
@@ -0,0 +1,177 @@
+        if (output == 'meilisearch retriever') {
+            return hybridsearchretriever
+        } else {
+            return hybridsearchretriever //TODO
```
is there any TODO item that we still need to do here?

MohamedYasserOaf replied:
yes , also like postgres, the component is intended to either return the documents retrieved with data type document , or you can return a baseRetriever to be used in the chain , but this is yet to be implemented, we can remove the extra output and only return the retriever for now and implement the other function in a future PR

HenryHengZJ commented on packages/components/nodes/retrievers/MeilisearchRetriever/MeilisearchRetriever.ts:
```diff
@@ -0,0 +1,177 @@
+
+    constructor() {
+        this.label = 'Meilisearch retriever'
+        this.name = 'Meilisearch retriever'
```
`name` has to be camelCase

HenryHengZJ commented on docker/Dockerfile-ntg:
```diff
@@ -0,0 +1,23 @@
+# Use the official Node.js image as the base image
```
can we remove all non PR related changes?

MohamedYasserOaf replied:
sure thing , do you want me to revert these changes in another commit on the same branch ? or do you want me to remove them from my branch ?

MohamedYasserOaf commented:
I want to change the branch name, the PR was closed when i did so, maybe when can change it when the comments are resolved
@HenryHengZJ

HenryHengZJ commented:
@MohamedYasserOaf I've updated the code, and moved it into Vector Store folder, can you give this a test?

MohamedYasserOaf commented:
I pulled the branch and applied pnpm install , the code fails on the build process , can't exactly point out the issue , i can send you the logs 
@HenryHengZJ

HenryHengZJ commented:
> I pulled the branch and applied pnpm install , the code fails on the build process , can't exactly point out the issue , i can send you the logs @HenryHengZJ

because this pulls in the main branch with 2.0 changes, you need to
1.) remove all the `node_modules`
2.) `pnpm i`
3.) `pnpm clean`
4.) `pnpm build-force`

MohamedYasserOaf commented:
@HenryHengZJ I removed all the node_modules and followed your instructions , the issue still persist , maybe because my main is not up to date ?

HenryHengZJ commented:
I was able to run it:
![image](https://github.com/user-attachments/assets/fb5560c6-49de-4658-a27d-136b93fc0ada)

but there is no documents upserted:
![image](https://github.com/user-attachments/assets/a053b23a-d5f5-4c9b-8d2c-10045ff3952a)

MohamedYasserOaf commented:
@HenryHengZJ my friend was able to run it too, I will check the dependencies from my side, you will find the documents upserted on your index but you will not get notified, this is a functionality i need to implement, maybe i can add it to this PR, you can try to chat with the flow to test in the meantime

HenryHengZJ commented:
no I dont see anything upserted on Meilisearch, and chatting with it always return nothing

MohamedYasserOaf commented:
I will lookup the issue after solving my build issue, we are already using our custom flowise image with meilisearch retriever in our production environment, providing great results, will check and come back to you

MohamedYasserOaf commented:
@HenryHengZJ Hello Henry, I resolved my dependency problems and re-installed everything, I rebased with the latest commits from the main repository, now the build step fails because of a component called Melvis, the same happens with my friend
https://github.com/FlowiseAI/Flowise/blob/7153093bfc8c30d7095ac747e9d6a43fbc804c5d/packages/components/nodes/vectorstores/Milvus/Milvus.ts#L250
the above line causes the problem
![image](https://github.com/user-attachments/assets/51dad4e7-d41b-41f4-af4e-c5d9407bfba2)
The code build when I apply @ts-ignore on this line 
![image](https://github.com/user-attachments/assets/dd5ff194-bf21-457d-ad97-bd24e3f64f22)

![image](https://github.com/user-attachments/assets/1888f3d2-871f-48d1-881f-263fb2e2ab07)
![image](https://github.com/user-attachments/assets/c8109284-7c11-4607-9a95-b2b6c90c82dd)
![image](https://github.com/user-attachments/assets/7f37cf27-a427-486f-b8f7-303f37a32255)
regardless of the poor answer and documents used (I run locally with no gpu and old processor), the component is up and running , and i also synced my forked and updated my branch

HenryHengZJ commented:
yeah I realized thhat as well and solved this in this [PR](https://github.com/FlowiseAI/Flowise/pull/2945/files#diff-5ed66dc721b47d417f6a058bb144a4a63e63dc45382e610f0582a76a41dc4e3d)

did you upsert the data on Flowise? or you are just using for retrieval?

MohamedYasserOaf commented:
yes I upserted using the flow, you tried doing so ? we use a UI to visualize our indexes، I can shoot you a small video or a series of screenshots of the process, we also added features like an option to delete an index, we added metadata as searchable attributes for appling various filters when retrieving.

HenryHengZJ commented:
ah okay because I was not able to do so when I try upserting, so I wonder did that work for you.

feel free to send me the screenshot/videos!

MohamedYasserOaf commented:
@HenryHengZJ Hello henry, the following screenshot shows the whole process,

This is a screenshot of our flow, the index used is "my-index-1" , test bool document store contains around 7 pdfs chunked using unstructured.io
![Screenshot from 2024-08-10 19-38-27](https://github.com/user-attachments/assets/77c50daf-bfca-47fc-9915-740bcbd67e02)

![image](https://github.com/user-attachments/assets/1564acf4-d2a2-416b-ba03-7899af1d6309)

These screenshots shows "my-index-1" having 0 documents
![Screenshot from 2024-08-10 19-19-10](https://github.com/user-attachments/assets/49837a63-961e-4828-951e-5772d72e4bb3)
![Screenshot from 2024-08-10 19-19-28](https://github.com/user-attachments/assets/95e0be90-5773-4c6d-8323-b2b7bf4a15a6)

The following screenshot shows my flowise upsert window
![image](https://github.com/user-attachments/assets/be28ad8d-69de-42a1-be3e-6376aa6f068b)

this is my index after upserting 
![Screenshot from 2024-08-10 19-42-19](https://github.com/user-attachments/assets/642ead50-3cd5-4c6a-8035-7fad32a6f143)
![image](https://github.com/user-attachments/assets/516a58d1-d204-4f6c-ac81-cd059dfbc64f)

This is a screenshot of my chat window using my flow, you can see it has retrieved context from my upserted index
![image](https://github.com/user-attachments/assets/8d0c4dc5-0f9f-4361-8341-c1fe7cc1fb9b)
![Screenshot from 2024-08-10 20-13-39](https://github.com/user-attachments/assets/ea5eef4b-ec31-4f6d-bf47-06c2e3660119)

- You can try to visualize your index in a 3rd party tools or via API calls to ensure your index is properly upserted 
- You want to make sure that you entered your admin api key, I encountered this at the beginning I was using only the search API key 
- Try to watch the terminal for any errors occured during upsert process

HenryHengZJ commented:
Thanks for the screenshot!
